### ISSUE TITLE
Doubt regarding incremental multi-task learning training 

### ISSUE THREAD

**Issue by prachigarg23** [menelaoskanakis/RCM#1]
Hi, thanks for your paper and code. I am working on Incremental multi-task learning for different fine-grained classification tasks. I found your research really useful and wish to baseline against it. Can you please tell me how the code can be used to train on the same model in an incremental manner? For example say that I want to incrementally learn on Edge detection followed by incrementally learn semantic segmentation followed by incrementally learning on human parts detection. So If I just run the training scripts with architectures having RCM modules sequentially, will that work?

**Comment by menelaoskanakis** [comment]
Hi. Thank you for your interest in our work. 

Referring to equation (2) in our paper, each convolution has been reparameterized with the RCM counterpart which is comprised of a non-trainable/frozen convolution (filter bank) and a task-specific convolution. Since the filter bank is frozen, when training RCM for any task (e.g. edge), you are actually only training the task-specific parameters for edge. More details on this can be found in Section 3.3.

You can therefore train the networks using the configs "semseg_RC_RI_NFF.yaml", "edge_RC_RI_NFF.yaml", etc. separately, and due to the disentanglement between the parameters, you end up with each model having the same filter bank parameters, and their own task-specific parameters.

Note: You can in principle train them in series (first train the model for edge, then use that model to initialize and train semseg etc), but since semseg cannot update the edge parameters this would not make any difference in the performance. As such, I implemented the training in the way mentioned above in order to allow for faster experimentation. 

I hope this helps clarify your question.

**Comment by prachigarg23** [comment]
Thanks for your response. I was just wondering how to add RCM layers to the same model if I want to sequentially incrementally learn on 2,3 tasks wherein the current task is trained by adding RCM layers on the previous task's initialisation and optimized on the current task's loss function. 

Also in the case of Deeplabv3+, I understand that every residual block in the encoder is reparameterized to contain a set of shared layers and domain specific layers. But I couldn't see any mention of what happens to the decoder and the classifier head if I want to implement a single incremental model. Would I need different decoder heads or classifier layers to be added for each different domain/task? 

Basically, how will the architecture look if a single model is being implemented for all the tasks? (like if a single model is being incrementally learnt on. I do understand that in your paper by construction it doesn't matter if you add new tasks to the same model or train different models for the ease of implementation.)

**Comment by menelaoskanakis** [comment]
When constructing the RCM module, it creates task-specific parameters for every task you would like to solve, and given the flag in the forward pass it activates the appropriate parameters. Currently, the codebase allows training of multiple tasks in a single model by backpropagating one task at a time (https://github.com/menelaoskanakis/RCM/blob/master/trainer.py#L150). If you would like to incrementally learn one task at a time, you would need to modify slightly the codebase. Specifically, initializing the model once at the beginning for all tasks, and then repeating the training loop of train.py (but hinting to trainer.py to only optimize a specific task instead of all the tasks as it currently does). In a simpler approach, you could just train one task at a time on its own script and then merge the checkpoints together (e.g. add-in the checkpoints of semseg the task-specific parameters of edge) although given your question this might not work for you. 

Our codebase currently uses task-specific decoders. The main reason is that the backbone already has been initialized with ImageNet weights and therefore has a meaningful filter bank, while the decoders are randomly initialized, so the filterbanks wouldn't make much sense. Note that this is also a common practice in most multi-task literature where all tasks share the same encoder but have task-specific decoders.

**Comment by prachigarg23** [comment]
Hi, thanks a lot for your explanations, I have a clear understanding of the work by now. 

w.r.t. "We additionally report the performance of a model trained jointly on all tasks, consisting of a fully shared encoder and task-specific decoders (Multi-task)." -  I was wondering if you could give some details on how you trained this joint model, if there is any convention on how to train a joint, multi-task or multi-domain model on different datasets. 

In my problem setting, I have a shared encoder and different decoder heads and I want to train a joint, task-aware multi-decoder model on the respective training datasets using respective task-specific losses of the different datasets being learnt on. Especially, can you please share how the dataloader should be written for such a joint learning experiment? I understand that once I have the task-specific loss from all decoder heads for all datasets, I can sum them and back propagate them at once, they will get updated in their respective decoders and all of them will get updated in the shared encoder. Any guidance will be appreciated.

**Comment by menelaoskanakis** [comment]
Sorry for the late response but I have been busy with other submissions and conference reviewing. 

There are a number of different ways this can be done. In our case, the datasets considered include labels for all the tasks, and therefore no special considerations are required (simple by sampling e.g. 8 images will include labeles for all the tasks). Due to this setup, one can easily forward propagate through each task-specific branch (think of for-loop over the tasks) and simply aggregate the losses without needing, for example, any custom pulling mechanisms for different datasets.

Since the additional questions seem unrelated to the codebase, but rather more generic guidance questions, please feel free to reach out to my email with more specifics of what exactly you are trying to do and I can try and provide you with specifics.

**Comment by prachigarg23** [comment]
Thanks, that helps a lot. Actually I found that in the file util.py > get_params() function, the learning rate for common branches was divided by the number of tasks being trained on the multi-task joint learning setting. That allows for easy forward propagation and backward propagation after aggregating the losses or iteratively updating the weights. So that answered my question. Thanks.
